### ISSUE TITLE
Context-aware mentions

### ISSUE THREAD

**Issue by fpesari** [Nheko-Reborn/nheko#1361]
### The Problem

Typing the `@` character brings up a seemingly random list of usernames, which in a group also includes some users who have never uttered a single word. So, IMHO it is not very meaningful.

### The Solution

`@` by default should bring up users which satisfy some criteria of relevance, which IMHO are in this order* ([1]):
1. Users who have recently posted something
2. Users who have mentioned you in the last N posts (where N can be configured)
3. Users who have quoted you in the last N posts (where N can be configured)
4. Special users (like admins and mods)

This should also happen when typing characters after the `@` following the `most recent` criteria. So, if the last 10 users who posted something in a group are
- Anna
- Burt: mentioned you
- Cecilia
- Derek: quoted you
- Ester: quoted you
- Frances
- Carl

- `@` should offer for example `Carl-Frances-Ester` (recent-first) or `Carl-Ester-Frances` (mixed, recent-quotes/mentions)
 `@C` should offer `Carl` as the first suggestion and `Cecilia` as the second suggestion.

* Actually, the order could be made configurable.

### Additional context

The Telegram desktop client seems to do something very similar to the solution I described.

### Happens in the latest version

- [X] Yes, this feature is still missing.

**Comment by deepbluev7** [comment]
The problem is more that then you can't early out after the first 7 matches, but instead have to iterate further, possibly all partial matches. Alternatively I need to figure out a better searching algorithm, that allows returning best matches ordered already to still allow only iterating a small space of possible completions.

**Comment by fpesari** [comment]
> The problem is more that then you can't early out after the first 7 matches, but instead have to iterate further, possibly all partial matches. Alternatively I need to figure out a better searching algorithm, that allows returning best matches ordered already to still allow only iterating a small space of possible completions.

I still did not have the time to look at Nheko's codebase but I add that, according to what I said:
1. You could first check if there is a non-empty cache. If there isn't, you can do things the way they are done right now
2. If there is a non-empty cache, then you could search those matches
3. If what the users searches for does not have any result in the cache, you can do things the way they are done right now
4. The search could initially adopt a `startsWith` criterion, then later move to a more flexible one like this: `@e` matches `Ester` even if `Cecilia` is more recent than `Cecilia`, but if there are no recent entries starting with `@e`, the first most recent username containing an `e` (`Cecilia`, in this case) is matched.

**Comment by deepbluev7** [comment]
No, specifically that we don't iterate all users would be a problem here. We build a completion trie once, when you open the completer the first time in a room. Then we iterate that trie using some permutation logic. To be able to easily rank stuff there, we would need to iterate way more elements and then sort them by ranking. That would significantly increase the search cost for short or empty queries. You can find most of the logic here: https://github.com/Nheko-Reborn/nheko/blob/master/src/CompletionProxyModel.h#L73

I would need to do some research into radix trees and stuff like that to figure out how others do that :)
